### PR TITLE
MultiToken transfer gas value calculation update

### PR DESCRIPTION
**Pull request by BonelessImpl**
## Summary by CodeRabbit

- **Refactor**
  - Improved efficiency of multi-token transfer and withdrawal operations by optimizing parameter passing and reducing unnecessary cloning.
  - Introduced dynamic gas calculation for multi-token operations, enhancing performance and reliability.
  - Modularized and reorganized trait implementations and imports in test utilities for better maintainability.
  - Reorganized import statements across multiple test files for clarity and reduced redundancy.

- **New Features**
  - Added new test utilities and traits for multi-token deposit and withdrawal operations.
  - Introduced comprehensive tests for multi-token enumeration, withdrawals, and gas usage analysis.
  - Added an iterator generating letter combinations for testing purposes.
  - Enhanced logging with detailed gas consumption reporting in test logs.
  - Added new transfer call variants supporting callback messages.

- **Bug Fixes**
  - Ensured accurate event emission and reduced redundant data handling in multi-token events.

- **Chores**
  - Added new development dependencies to the test suite for improved testing capabilities.

- **Tests**
  - Expanded and reorganized test coverage for multi-token functionality, including edge cases and performance under varying conditions.
  - Removed obsolete tests and replaced with updated modular test implementations.

### CODE REVIEW

**Review by fusede — CHANGES_REQUESTED** [review]

File: defuse/src/contract/gas.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+const MT_RESOLVE_TRANSFER_GAS_PER_TOKEN: Gas = Gas::from_tgas(14);
+const MT_RESOLVE_TRANSFER_BASE_GAS: Gas = Gas::from_tgas(5);
+
+#[must_use]
+pub fn total_mt_withdraw_gas(token_count: usize) -> Gas {
+    let token_count: u64 = token_count.try_into().unwrap_or_panic_display();
+
+    MT_RESOLVE_TRANSFER_BASE_GAS
+        .saturating_add(MT_RESOLVE_TRANSFER_GAS_PER_TOKEN.saturating_mul(token_count))
+}
```
[the remark this answers is not in the document]
Comment (BonelessImpl) [reply defuse/src/contract/gas.rs:24]: The reason this is public is because it's used in the tests.

The values are exaggerated to account for risk/errors. Feel free to suggest new values based on the calculations I made. This is all about being comfortable with future changes. 5 seems too much, agreed. But I don't think it matters that much because it doesn't scale with the number of tokens.

File: tests/src/tests/defuse/tokens/nep141/traits.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
+use defuse::tokens::DepositMessage;
```
[the remark this answers is not in the document]
Comment (BonelessImpl) [reply tests/src/tests/defuse/tokens/nep141/traits.rs:1]: I disagree. But regardless, can you tell me what you expect to happen here? What would you like to call this module?

File: tests/src/utils/test_log.rs [hunk 3]
```diff
@@ -1,10 +1,13 @@
+use near_sdk::Gas;
 use near_workspaces::result::ExecutionResult;
 
 #[allow(dead_code)]
 #[derive(Debug)]
 pub struct TestLog {
     logs: Vec<String>,
     receipt_failure_errors: Vec<String>,
+    gas_burnt_in_tx: Gas,
```
[the remark this answers is not in the document]
Comment (BonelessImpl) [reply tests/src/utils/test_log.rs:9]: This does use execution result. This struct is just simplifies extracting more data from ExecutionResult because it can be at times obnoxious. Check the `From` implementation below.

[the remark this answers is not in the document]
Comment (fusede) [reply tests/src/utils/test_log.rs:9]: I've found it lacking some info in [#102](https://github.com/near/intents/pull/102/files#diff-25958c7fef7e3624b66c27dc56698b4180edd9130fe7de3a649dfb70b488f8e5R72), so I ended up with just printing ExecutionResult.
This struct will eventually grow and become a clone of `ExecutionResult`

File: defuse/src/contract/tokens/nep245/withdraw.rs [hunk 4]
```diff
@@ -20,10 +24,20 @@ use near_sdk::{
...
 
-use crate::{
-    contract::{Contract, ContractExt, Role, tokens::STORAGE_DEPOSIT_GAS},
-    tokens::nep245::{MultiTokenForceWithdrawer, MultiTokenWithdrawResolver, MultiTokenWithdrawer},
-};
+// These represent a linear model total_gas_cost = per_token*n + base,
+// where `n` is the number of tokens in one mt_withdraw call.
+// fitted from points {(1, 3.0), (2, 3.0), (3, 3.0)}
+// The slope is below the margin of error. So we choose a low enough value.
+const MT_RESOLVE_WITHDRAW_PER_TOKEN_GAS: Gas = Gas::from_tgas(1);
```
Comment (fusede) [inline defuse/src/contract/tokens/nep245/withdraw.rs:31]: This should be similar to `MT_RESOLVE_TRANSFER_GAS`, set to 2
And base to 8

File: defuse/src/contract/tokens/nep245/withdraw.rs [hunk 5]
```diff
@@ -124,12 +122,22 @@ impl Contract {
         )
         .into())
     }
+
+    #[must_use]
+    fn mt_withdraw_resolve_gas(token_count: usize) -> Gas {
+        // Values chosen to be similar to `MT_RESOLVE_TRANSFER_*` values
+        const MT_RESOLVE_WITHDRAW_PER_TOKEN_GAS: Gas = Gas::from_tgas(2);
+        const MT_RESOLVE_WITHDRAW_BASE_GAS: Gas = Gas::from_tgas(8);
+
```
Comment (BonelessImpl) [inline defuse/src/contract/tokens/nep245/withdraw.rs:131]: Shouldn't we just use the same constants from `mt_resolve_gas()`, like make the constants public and use them? This will help in case we decide to change them in the future.

Comment (fusede) [reply defuse/src/contract/tokens/nep245/withdraw.rs:131]: I'd better not... Despite they have same values, they are still intended for different Promises

File: defuse/src/contract/tokens/nep245/core.rs [hunk 6]
```diff
@@ -238,15 +238,19 @@ impl Contract {
...
         const MT_RESOLVE_TRANSFER_PER_TOKEN_GAS: Gas = Gas::from_tgas(2);
         const MT_RESOLVE_TRANSFER_BASE_GAS: Gas = Gas::from_tgas(8);
         let token_count: u64 = token_count.try_into().unwrap_or_panic_display();
 
         MT_RESOLVE_TRANSFER_BASE_GAS
-            .checked_add(MT_RESOLVE_TRANSFER_PER_TOKEN_GAS.saturating_mul(token_count))
+            .checked_add(
+                MT_RESOLVE_TRANSFER_PER_TOKEN_GAS
+                    .checked_mul(token_count)
+                    .unwrap(),
```
Comment (fusede) [inline defuse/src/contract/tokens/nep245/core.rs:252]:
```suggestion
                    .unwrap_or_panic(),
```

**Review by fusede — APPROVED** [review]
